### PR TITLE
Add NATS JetStream scaler

### LINKED ISSUE #2391: Provide scaler for NATS Jetstream
### Proposal

Currently keda supports NATS streaming. However, [NATS  Streaming has been deprecated in favor of NATS Jetstream](https://docs.nats.io/reference/faq#jetstream-and-nats-streaming). NATS is simpler to run in-cluster than Kafka, and jetstream scales well. I hope that Keda will support it?

### Use-Case

I hope to use in [conjunction with fission](https://github.com/fission/fission/issues/2062)

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by goku321**
Add NATS JetStream scaler

### Checklist

- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))
- [x] Tests have been added
- [N/A] A PR is opened to update our Helm chart ([repo](https://github.com/kedacore/charts)) *(if applicable, ie. when deployment manifests are modified)*
- [x] A PR is opened to update the documentation on ([repo](https://github.com/kedacore/keda-docs/pull/881)) *(if applicable)*
- [x] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)

Fixes #2391 

Documentation https://github.com/kedacore/keda-docs/pull/881

### CODE REVIEW

**Comment by goku321** [comment]
I tested this scaler using kind and could only test with `maxReplicaCount: 2` due to limited hardware resources. Also, will create a docs PR soon.

File: pkg/scalers/nats_jetstream_scaler.go
Comment (zroubalik) [inline pkg/scalers/nats_jetstream_scaler.go:11]: please put KEDA related imports to the separate block below

File: pkg/scalers/nats_jetstream_scaler.go
Comment (zroubalik) [inline pkg/scalers/nats_jetstream_scaler.go:148]: We have recently introduced configurable activation, could you please add it here: https://github.com/kedacore/keda/issues/2800

File: tests/scalers/nats-jetstream-helpers.ts
Comment (zroubalik) [inline tests/scalers/nats-jetstream-helpers.ts:1]: We are finishing migration of e2e tests to Go, could you please rewrite this?

https://github.com/kedacore/keda/issues/2737

File: CHANGELOG.md
Comment (zroubalik) [inline CHANGELOG.md:45]:
```suggestion
- **General:** Add new NATS JetStream scaler ([#2391](https://github.com/kedacore/keda/issues/2391))
```

**Comment by zroubalik** [comment]
@goku321 any update on this please?

**Comment by goku321** [comment]
@zroubalik I will work on the changes today/tomorrow and make it ready to be merged :)

**Comment by zroubalik** [comment]
> @zroubalik I will work on the changes today/tomorrow and make it ready to be merged :)

@goku321 any updates please? we would need to merge it today in order to get it into the next release, if not that we will target 2.9.

**Comment by goku321** [comment]
Hi @zroubalik Going to push the changes very soon. Converting tests to go took some time and running tests locally in kind is a bit slower 😅

File: CHANGELOG.md
Comment (zroubalik) [inline CHANGELOG.md:46]: Please, sort alphabetically.

File: pkg/scalers/nats_jetstream_scaler.go
Comment (zroubalik) [inline pkg/scalers/nats_jetstream_scaler.go:169]:
```suggestion
		jsLog.Error(err, "unable to access NATS JetStream monitoring endpoint", "natsServerMonitoringEndpoint", s.metadata.monitoringEndpoint)
```

Comment (zroubalik) [reply pkg/scalers/nats_jetstream_scaler.go:169]: please use lowercase at the beginning of error messages, the same apply for the other error messages in this scaler.

File: tests/scalers/nats/nats_jetstream/nats_jetstream_test.go
Comment (zroubalik) [inline tests/scalers/nats/nats_jetstream/nats_jetstream_test.go:64]: could you please add source code for this image here: https://github.com/kedacore/test-tools/tree/main/e2e/images ?

**Review by zroubalik — COMMENTED** [review]
looking good, just a couple of minor nits.

**Comment by zroubalik** [comment]
@goku321 also please fix the static checks: https://github.com/kedacore/keda/runs/7720862248?check_suite_focus=true

File: tests/scalers/nats/nats_jetstream/nats_jetstream_test.go
Comment (zroubalik) [reply tests/scalers/nats/nats_jetstream/nats_jetstream_test.go:64]: (This could be done after this PR is merged)

File: tests/scalers/nats/nats_jetstream/nats_jetstream_test.go
Comment (zroubalik) [inline tests/scalers/nats/nats_jetstream/nats_jetstream_test.go:4]: the package should be equal to the directory name

File: tests/scalers/nats/helper/helper.go
Comment (zroubalik) [inline tests/scalers/nats/helper/helper.go:4]: please follow a pattern for file/package naming of a scaler that's used for example here:
https://github.com/kedacore/keda/tree/main/tests/scalers/prometheus

**Comment by zroubalik** [comment]
@goku321 could you please rebase your branch agains main? There has been some updates recently (we haver removed npm) that are failing here.

**Comment by zroubalik** [comment]
@goku321 also linter is failing, you can check it locally by running: `make golangci`

**Comment by goku321** [comment]
@zroubalik linter is complaining for duplicate code b/w stan and nats_jetstream scaler. Some part is overlapping

**Comment by zroubalik** [comment]
> @zroubalik linter is complaining for duplicate code b/w stan and nats_jetstream scaler. Some part is overlapping

In that case please add exception for your scaler to this file: https://github.com/kedacore/keda/blob/acb03ab4daf53fa82c00be2534300fef65f30eb4/.golangci.yml#L67

Also, plase squash the commits, so there are only relevant commits in this PR, otherwise the DCO will complain.

**Comment by goku321** [comment]
@zroubalik I merged from main into this branch a few times instead of rebasing so I'm not sure if it is possible to squash now. Possible to do "squash and merge" from the drop down while merging?

**Comment by zroubalik** [comment]
@goku321 yeah, I can squash those commits. I just don't want to have any unrelated commits here.

**Comment by goku321** [comment]
@zroubalik Also, I updated `.golangci.yml` for "nats_jetstream_scaler". Do I need to add a path for "stan_scaler" as well?

**Comment by zroubalik** [comment]
> @zroubalik Also, I updated `.golangci.yml` for "nats_jetstream_scaler". Do I need to add a path for "stan_scaler" as well?

I am not sure, but given the error it seems like that you'd have to do that :(

Also please fix the package name in the e2e test, see the other linter error. Then we are hopefully good to test this scaler here in the PR and the only outstanding issue will be docs PR 🤞

**Review by zroubalik — APPROVED** [review]
LGTM

**Comment by zroubalik** [comment]
@goku321 just the Documentation is missing.

**Comment by goku321** [comment]
@zroubalik Working on the documentation 🙂

**Comment by zroubalik** [comment]
> @zroubalik Working on the documentation 🙂

@goku321 what is the status? If we don't merge it today it won't get into the release unfortunately

**Comment by goku321** [comment]
@zroubalik Added the documentation. Thanks for your patience 😅

**Comment by zroubalik** [comment]
@goku321 I have just realized that the PR is missing a Unit test. I completely missed this somehow.

Could you please add one, it's quite an easy task, if you look at other scalers.

**Comment by goku321** [comment]
@zroubalik I had that in back of my mind but forgot about it. Will add for sure

**Pull request merged**
